cli/src/lib/uploadManager.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  Operation,
+  OperationStatus,
+  SmartContract,
+} from '@massalabs/massa-web3'
+
+import { Batch } from '../lib/batcher'
+import { uploadChunks } from '../lib/website/uploadChunk'
+
+export interface UploadManagerOptions {
+  maxConcurrentOps: number
+}
+
+export enum BatchStatus {
+  WaitingUpload = 0,
+  Sent = 1,
+  Success = 2,
+  Error = 3,
+}
+
+export interface UploadBatch extends Batch {
+  operation?: Operation
+  status: BatchStatus
+  gas: bigint
+}
+
+export class UploadManager {
+  private batches: UploadBatch[]
+  private maxConcurrentOps: number
+  private operations: Operation[] = []
+
+  constructor(
+    batchedChunks: UploadBatch[],
+    options: UploadManagerOptions = { maxConcurrentOps: 1 }
+  ) {
+    this.batches = batchedChunks.map((batch, index) => ({
+      ...batch,
+      id: index,
+      status: BatchStatus.WaitingUpload,
+    }))
+    this.maxConcurrentOps = options.maxConcurrentOps
+  }
+
+  async startUpload(
+    sc: SmartContract,
+    onUpdate: (batch: UploadBatch) => void
+  ): Promise<void> {
+    const activeUploads: Promise<void>[] = []
+
+    for (const batch of this.batches) {
+      if (activeUploads.length >= this.maxConcurrentOps) {
+        await Promise.race(activeUploads)
+      }
```

Comment:
For a future version, we could potentially improve efficiency by starting a new upload immediately when one finishes, while still respecting the maxConcurrentOps limit

Reply:
I don't understand, that's already what we do 🤔